src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -239,62 +411,108 @@ ucc_tl_ucp_reduce_scatter_knomial_finalize(ucc_coll_task_t *coll_task)
+    size_t                max_seg   = task->reduce_scatter_kn.max_seg;
+    size_t data_size;
+    ucc_kn_radix_t step_radix;
+    size_t max_recv_size;
+
+    if (args->coll_type == UCC_COLL_TYPE_ALLREDUCE) {
+        if (KN_NODE_EXTRA != task->reduce_scatter_kn.p.node_type) {
+            if (coll_args->mask & UCC_BASE_CARGS_MAX_FRAG_COUNT) {
+                count = coll_args->max_frag_count;
+            }
+            data_size = count * dt_size;
+            step_radix = ucc_kn_compute_step_radix(&task->reduce_scatter_kn.p);
+            max_recv_size = ucc_sra_kn_compute_seg_size(count, step_radix, 0) *
+                step_radix * dt_size;
+
+            if (UCC_IS_INPLACE(coll_args->args) ||
+                (KN_NODE_PROXY == task->reduce_scatter_kn.p.node_type) ||
+                max_recv_size > data_size) {
+                return ucc_max(max_recv_size, data_size);
+            } else {
+                return 0;
+            }
+        }
+    } else {
+        step_radix = task->reduce_scatter_kn.p.radix;
+        if (KN_NODE_PROXY == task->reduce_scatter_kn.p.node_type) {
+            if (UCC_IS_INPLACE(*args)) {
+                return max_seg * step_radix * dt_size;
+            } else {
+                return (max_seg * (step_radix - 1) + count) * dt_size;
+            }
+        } else if (KN_NODE_EXTRA != task->reduce_scatter_kn.p.node_type) {
+            return max_seg * step_radix * dt_size;
+        }
+    }
+    return 0;
+}
+
 ucc_status_t
 ucc_tl_ucp_reduce_scatter_knomial_init_r(ucc_base_coll_args_t *coll_args,
                                          ucc_base_team_t *team,
                                          ucc_coll_task_t **task_h,
                                          ucc_kn_radix_t radix)
 {
     ucc_tl_ucp_team_t *tl_team   = ucc_derived_of(team, ucc_tl_ucp_team_t);
-    size_t             count     = coll_args->args.dst.info.count;
-    ucc_datatype_t     dt        = coll_args->args.dst.info.datatype;
-    size_t             dt_size   = ucc_dt_size(dt);
     ucc_memory_type_t  mem_type  = coll_args->args.dst.info.mem_type;
+    size_t             count     = GET_COUNT(&coll_args->args);
+    ucc_coll_type_t    ct        = coll_args->args.coll_type;
     ucc_sbgp_t        *sbgp;
     ucc_tl_ucp_task_t *task;
     ucc_status_t       status;
-    size_t             max_recv_size, data_size;
-    ucc_kn_radix_t     step_radix;
+    size_t             scratch_size;
     ucc_rank_t         rank, size;
+    ptrdiff_t          max_seg_offset;
 
     task                 = ucc_tl_ucp_init_task(coll_args, team);
     task->super.flags    |= UCC_COLL_TASK_FLAG_EXECUTOR;
     task->super.post     = ucc_tl_ucp_reduce_scatter_knomial_start;
     task->super.progress = ucc_tl_ucp_reduce_scatter_knomial_progress;
     task->super.finalize = ucc_tl_ucp_reduce_scatter_knomial_finalize;
 
-    if (tl_team->cfg.use_reordering) {
+    if (tl_team->cfg.use_reordering && ct == UCC_COLL_TYPE_ALLREDUCE) {
```

Comment:
I am not sure to understand. In what case is it allreduce here?

Reply:
SRA allreduce in UCC is a combination of reduce scatter knomial and allgather knomial. Depending on whether the reduce scatter is standalone coll or part of SRA allreduce we need sometimes to make different decisions.